Userland/Libraries/LibGfx/ImageFormats/AVIFLoader.cpp

```diff
@@ -0,0 +1,199 @@
+        auto bitmap_format = context.has_alpha ? BitmapFormat::BGRA8888 : BitmapFormat::BGRx8888;
+        auto bitmap = TRY(Bitmap::create(bitmap_format, context.size.value()));
+
+        avifRGBImageSetDefaults(&rgb, context.decoder->image);
+        rgb.pixels = bitmap->scanline_u8(0);
+        rgb.rowBytes = bitmap->pitch();
+        rgb.format = avifRGBFormat::AVIF_RGB_FORMAT_BGRA;
+
+        avifResult result = avifImageYUVToRGB(context.decoder->image, &rgb);
+        if (result != AVIF_RESULT_OK)
+            return Error::from_string_literal("Conversion from YUV to RGB failed");
+
+        auto duration = context.decoder->imageCount == 1 ? 0 : static_cast<int>(context.decoder->imageTiming.duration * 1000);
+        context.frame_descriptors.append(ImageFrameDescriptor { bitmap, duration });
+
+        context.state = AVIFLoadingContext::BitmapDecoded;
+    }
+
+    return {};
+}
+
+IntSize AVIFImageDecoderPlugin::size()
+{
+    return m_context->size.value();
+}
+
+bool AVIFImageDecoderPlugin::sniff(ReadonlyBytes data)
+{
+    AVIFLoadingContext context;
+    context.data = data;
+    return !decode_avif_header(context).is_error();
+}
+
+ErrorOr<NonnullOwnPtr<ImageDecoderPlugin>> AVIFImageDecoderPlugin::create(ReadonlyBytes data)
+{
+    auto context = TRY(try_make<AVIFLoadingContext>());
+    auto plugin = TRY(adopt_nonnull_own_or_enomem(new (nothrow) AVIFImageDecoderPlugin(data, move(context))));
+    TRY(decode_avif_header(*plugin->m_context));
+    return plugin;
+}
+
+bool AVIFImageDecoderPlugin::is_animated()
+{
+    return m_context->image_count > 1;
+}
+
+size_t AVIFImageDecoderPlugin::loop_count()
+{
+    return is_animated() ? m_context->repetition_count : 0;
+}
+
+size_t AVIFImageDecoderPlugin::frame_count()
+{
+    if (!is_animated())
+        return 1;
+    return m_context->image_count;
+}
+
+size_t AVIFImageDecoderPlugin::first_animated_frame_index()
+{
+    return 0;
+}
+
+ErrorOr<ImageFrameDescriptor> AVIFImageDecoderPlugin::frame(size_t index, Optional<IntSize>)
+{
+    if (index >= frame_count())
+        return Error::from_string_literal("AVIFImageDecoderPlugin: Invalid frame index");
+
+    if (m_context->state == AVIFLoadingContext::State::Error)
+        return Error::from_string_literal("AVIFImageDecoderPlugin: Decoding failed");
+
+    if (m_context->state < AVIFLoadingContext::State::BitmapDecoded) {
+        TRY(decode_avif_image(*m_context));
+        m_context->state = AVIFLoadingContext::State::BitmapDecoded;
+    }
+
+    if (index >= m_context->frame_descriptors.size())
+        return Error::from_string_literal("AVIFImageDecoderPlugin: Invalid frame index");
+    return m_context->frame_descriptors[index];
+}
+
+ErrorOr<Optional<ReadonlyBytes>> AVIFImageDecoderPlugin::icc_data()
+{
+    if (m_context->state < AVIFLoadingContext::State::HeaderDecoded)
+        (void)frame(0);
+
+    if (!m_context->icc_data.is_empty())
+        return m_context->icc_data;
+    return OptionalNone {};
```

Comment:
`return m_context->icc_data` should work here

Reply:
I'd like to explictly return None though, right? With this it seems to return the empty vector.

Reply:
It seems like it's unnecessary for the API to return Optional, if a null or empty Span is also valid. :thinking:

Reply:
I agree. Should we go ahead with an API change to the `ImagePlugin` in this change or separately?

Reply:
Let's not touch the API in this PR.

I thought that `icc_data` was already an Optional, my bad. This was correct.